### PR TITLE
pySPM derived pixel scaling

### LINKED ISSUE #121: Switch to pySPM derived pixel to nanometere scaling factor
**Describe the bug**
Currently the workflow uses a parameter `dx` defined in the configuration that defines the pixel to nanometer scaling.

However, this can be extracted from the metadata of a `.spm` image by methods available in the `pySPM` package.

The function `extract_pixel_to_nanometer()` has been defined to do so, but switching to this will break tests from the point the parameter `dx` is used in finding grains and so the switch has not yet been made, but it should be changed and all tests updated in light of this.

### PR DESCRIPTION
**Pull request by ns-rse**
Resolves #121 

Switches to using the pySPM metadata metrics for the pixel to nanometer scaling which @SylviaWhittle informed me about when chatting the other day.

One impact is that two less grains are identified, but the issue of what a minimum grain size should be is still open for discussion and I have some reservations about defining it as median - 1.5 * Inter-quartile Range which I'll try and articulate with some simulations before meeting on Thursday.  @SylviaWhittle has already highlighted that such an approach on the raw data has potential to result in a minimum size that is negative, which is currently addressed using the Otsu threshold method to remove very small artefacts prior to setting a minimum threshold but this may not generalise well, particularly with say silicone samples rather than molecules.

### CODE REVIEW

**Comment by SylviaWhittle** [comment]
Thanks for investigating this @ns-rse. Do you think the minimum threshold should be set exclusively by the user since whatever decision we make on how to automatically detect a lower size limit will affect the output drastically?

**Comment by bobturneruk** [comment]
I will look at the code, but the matter of thresholding will need consent from the wider team.

**Comment by bobturneruk** [comment]
Should `find_grains.py` be called `grains.py`? `find_grains.py` makes me think it's going to find grains rather than be a collection of functions used in the finding of grains, which is what it is.

File: topostats/find_grains.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def quadratic(x, a, b, c):
 
 def gaussian_filter(image: np.array,
                     gaussian_size: float = 2,
-                    dx: float = 1,
+                    pixel_nm_scaling: float = 1,
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply topostats/find_grains.py:31]: Removed defaults for both `pixel_n_scaling` and `gaussian_size` they are explicitly stated when the function is called from `topotracing.py` and the values are from the config file.

File: topostats/find_grains.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def gaussian_filter(image: np.array,
     FIXME: Add docs.
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply topostats/find_grains.py:54]: Everything needs examples, but most methods are unlikely to be called/accessed directly so I'm not sure of the utility of adding them, for now I've removed them (they're mostly an artefact of me using the Emacs [numpy docstring plugin for Emacs](https://ddavis.io/posts/numpydoc-dot-el/) and not removing the unused `Example` section.

File: topostats/topotracing.py [hunk 3]
```diff
@@ -165,9 +166,13 @@ def process_scan(image_path: Union[str, Path] = None,
     tidied_borders = tidy_border(boolean_image_mask)
     LOGGER.info(f'[{img_name}] : Borders tidied.')
 
+    # Calculate minimum grain size (using pySPM derived pixel_nm_scaling)
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply topostats/topotracing.py]: Tardiness on my behalf not tidying up the comment, its in pixels, comment now tidied up. The pixel to nm scaling (now derived from

File: config/example.yaml [hunk 4]
```diff
@@ -9,7 +9,6 @@ channel: Height
 amplify_level: 1
 grains:
   gaussian_size: 2.0
-  dx: 1.0
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply config/example.yaml:12]: There is #123 for updating the documentation in light of all changes. Not sure what else might change so would like to re-write it in one go.

[the remark this answers is not in the document]
Comment (bobturneruk) [reply config/example.yaml:12]: I think #123 is enough, but we'll need to spend some time on this.

File: topostats/find_grains.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bobturneruk) [reply topostats/find_grains.py]: Great.

**Comment by ns-rse** [comment]
> Should `find_grains.py` be called `grains.py`? `find_grains.py` makes me think it's going to find grains rather than be a collection of functions used in the finding of grains, which is what it is.

I'm not too beholden to names and happy to change `find_grains.py` > `grains.py` and have just done so for #136. One thing I have in the pipeline is a class for `Filters` and a class for `Grains`. They're mostly written on a separate branch (`ns-rse/90-refactor-filters`) but I'm going to hold back on #133 being merged to `dev` then make a clean fork and get everything working using the tests in place there.

**Comment by ns-rse** [comment]
> I think there was some chat about moving lines 18-99 of `filters.py` to an separate place for handling interactions with files? Maybe out of scope for this PR. Is there an issue for this? Should there be?

After the discussion yesterday and some thought whilst running this morning I think this would be a good idea as a separate issue and the way to handle it would be via a [factory method](https://realpython.com/factory-method-python/) as I have done for thresholding in the pending PR #119. `process_scan()` would then call a function e.g. `load_scan()` with the file extension as an argument and the appropriate method would be used to return the data.

It would be useful to develop this to have another file format.  I'll create an issue for this shortly.

**Comment by ns-rse** [comment]
> I will look at the code, but the matter of thresholding will need consent from the wider team.

I have some thoughts on this as I don't think it makes sense to use the median -/+ <factor> * Inter-Quartile Range.

Median and Mean are both measures of centrality on distributions, for Normal/Gaussian distributions they are typically very similar. But each have their own way of quantifying variation around them, the Inter Quartile Range for the Median and the Standard Deviation for the mean.

There are methods such as [Winsorisation](https://en.wikipedia.org/wiki/Winsorizing) that are used for handling outliers when analysing data. Clearly we're not going to replace the size of such grains, but simply saying that you want to cull grains below the 2.5%-tile or above the 97.5%-tile could be applied.

Alternatively, if after removing all really small "noise" from a picture you have a roughly Gaussian/Normal distribution of grains sizes then similar could be achieved by taking the mean -/+ 1.96 x Standard Deviation which will typically capture the central 95% of a Gaussian/Normal distribution.

**Pull request merged**